scheduler/reconnecting_picker/reconnecting_picker.go

```diff
@@ -0,0 +1,108 @@
+	replacementIsNewer := replacement.Job.Version > original.Job.Version ||
+		replacement.Job.CreateIndex > original.Job.CreateIndex
+	if replacementIsNewer {
+		rp.logger.Debug("replacement has a newer version, keeping replacement")
+		return replacement
+	}
+
+	var picker func(*structs.Allocation, *structs.Allocation) *structs.Allocation
+
+	rs := ds.ReconcileStrategy()
+	rp.logger.Debug("picking according to strategy", "strategy", rs)
+
+	switch rs {
+	case structs.ReconcileOptionBestScore:
+		picker = rp.pickBestScore
+
+	case structs.ReconcileOptionKeepOriginal:
+		picker = rp.pickOriginal
+
+	case structs.ReconcileOptionKeepReplacement:
+		picker = rp.pickReplacement
+
+	case structs.ReconcileOptionLongestRunning:
+		picker = rp.pickLongestRunning
+	}
+
+	return picker(original, replacement)
+}
+
+// pickReconnectingAlloc returns the allocation to keep between the original
+// one that is reconnecting and one of its replacements.
+//
+// This function is not commutative, meaning that pickReconnectingAlloc(A, B)
+// is not the same as pickReconnectingAlloc(B, A). Preference is given to keep
+// the original allocation when possible.
+func (rp *ReconnectingPicker) pickBestScore(original *structs.Allocation, replacement *structs.Allocation) *structs.Allocation {
+
+	// Check if the replacement has better placement score.
+	// If any of the scores is not available, only pick the replacement if
+	// itself does have scores.
+	originalMaxScoreMeta := original.Metrics.MaxNormScore()
+	replacementMaxScoreMeta := replacement.Metrics.MaxNormScore()
+
+	replacementHasBetterScore := originalMaxScoreMeta == nil && replacementMaxScoreMeta != nil ||
+		(originalMaxScoreMeta != nil && replacementMaxScoreMeta != nil &&
+			replacementMaxScoreMeta.NormScore > originalMaxScoreMeta.NormScore)
+
+	// Check if the replacement has better client status.
+	// Even with a better placement score make sure we don't replace a running
+	// allocation with one that is not.
+	replacementIsRunning := replacement.ClientStatus == structs.AllocClientStatusRunning
+	originalNotRunning := original.ClientStatus != structs.AllocClientStatusRunning
+
+	if replacementHasBetterScore && (replacementIsRunning || originalNotRunning) {
+		return replacement
+	}
+
+	return original
+}
+
+func (rp *ReconnectingPicker) pickOriginal(original *structs.Allocation, _ *structs.Allocation) *structs.Allocation {
+	return original
+}
+
+func (rp *ReconnectingPicker) pickReplacement(_ *structs.Allocation, replacement *structs.Allocation) *structs.Allocation {
+	return replacement
+}
+
+func (rp *ReconnectingPicker) pickLongestRunning(original *structs.Allocation, replacement *structs.Allocation) *structs.Allocation {
+	lt := original.GetLeaderTask()
+
+	// Default to the first task in the group if no leader is found.
+	if lt.Name == "" {
+		lt = *original.Job.LookupTaskGroup(original.TaskGroup).Tasks[0]
+	}
+
+	// If the replacement has a later start time, keep the original.
```

Comment:
Technically it can happen that the LatestStartOfTask returns Zero, but it makes no sense in reality, should we address these cases?

Reply:
I think you can hit that case if the allocation has started but the main task hasn't yet when we disconnect and reconnect. So for example, if there's a long-running pre-start task. So I think we need to account for that.